gpytorch/kernels/kernel.py

```diff
@@ -193,29 +192,28 @@ def __pdist_dist(self, x1):
 
         return res
 
-    def __slow_sq_dist(self, x1, x2, diag, x1_eq_x2):
+    @torch.jit.script
+    def __jit_sq_dist(x1, x2, diag, x1_eq_x2):
         # Compute squared distance matrix using quadratic expansion
         x1_norm = x1.pow(2).sum(dim=-1, keepdim=True)
-        x2_norm = x2.pow(2).sum(dim=-1, keepdim=True)
-
-        if diag:
-            mid = (x1 * x2).sum(dim=-1, keepdim=True)
-            res = (x1_norm - 2 * mid + x2_norm).squeeze(-1)
+        if bool(x1_eq_x2):
+            x2_norm = x1_norm
         else:
-            mid = x1.matmul(x2.transpose(-2, -1))
-            res = x1_norm - 2 * mid + x2_norm.transpose(-2, -1)
+            x2_norm = x2.pow(2).sum(dim=-1, keepdim=True)
 
-        if x1_eq_x2:
+        res = x1.matmul(x2.transpose(-2, -1))
+        res = res.mul_(-2).add_(x2_norm.transpose(-2, -1)).add_(x1_norm)
+
+        if bool(x1_eq_x2):
             # Ensure zero diagonal
-            diag_inds = torch.arange(x1.shape[-2])
-            res[..., diag_inds, diag_inds] = 0
+            res.diagonal(dim1=-2, dim2=-1).fill_(0)
 
         # Zero out negative values
         res.clamp_min_(0)
 
         return res
 
-    def _covar_dist(self, x1, x2, diag=False, batch_dims=None, square_dist=False, **params):
+    def _covar_dist(self, x1, x2, diag=False, batch_dims=None, square_dist=False, jit_func=None, **params):
         """
         This is a helper method for computing the Euclidean distance between
         all pairs of points in x1 and x2.
```

Comment:
maybe explain in the docstring what `jit_func` is and how it's supposed to be used?

Reply:
I don't see why we'd ever not use the JIT version. The `jit_func` isn't intended to provide a squared distance function, rather it overrides what `_covar_dist` does with custom behavior. This is done because for RBF specifically, we want the exp() and div() done in the JIT script because it saves additional memory.